### ISSUE TITLE
URI module: UnicodeDecodeError in error callback

### ISSUE THREAD

**Issue by hryamzik** [ansible/ansible-modules-core#2166]
I have a simple task with URI module that fails:

``` yml
- name: get sub.class1.server.ca.pem
  uri: url=http://www.startssl.com/certs/sub.class1.server.ca.pem return_content=yes
  register: result
```

Here's the output:

```
Traceback (most recent call last):
  File "<stdin>", line 2081, in <module>
  File "<stdin>", line 447, in main
TypeError: don't know how to handle UnicodeDecodeError in error callback
```

Task fails on Centos 6 and SmartOS with ansible 1.9.3, worked a while ago on Ubuntu with ansible 1.9.2.

**Comment by robynbergeron** [comment]
[module: network/basics/uri.py]

**Comment by robynbergeron** [comment]
Hi @hryamzik -- thanks for submitting this bug report. 

@romeotheriault --  thanks in advance for taking a look at this bug report and for your ongoing work in maintaining this module.

**Comment by robynbergeron** [comment]
Also pinging @privateip for networking-ish-ness eyes :)

**Comment by hoxu** [comment]
I get this same issue with ansible 1.9.4 uri module:

```
TypeError: don't know how to handle UnicodeDecodeError in error callback
```

Here is the relevant task:

```
uri: url={{ url }} dest={{ filename }} validate_certs=no force_basic_auth=yes user={{ username }} password={{ password }}
```

With ansible 1.9.2 the same works.

**Comment by bcoca** [comment]
callbacks are not related at all to which module you execute, I doubt that this is a uri issue.

**Comment by hryamzik** [comment]
@hoxu just switch to "shell: curl"

@bcoca so this URL: `http://www.startssl.com/certs/sub.class1.server.ca.pem` is simply wrong? How can I fix it?

**Comment by hoxu** [comment]
See my comment above - the same task and URL works with 1.9.2, but not with 1.9.4. So the `UnicodeDecodeError` exception must be caused by some regression in ansible.

**Comment by abadger** [comment]
I think @bcoca means that the problem is probably in the callback portion of ansible rather than the uri module.

**Comment by abadger** [comment]
Does not traceback in v2.  However, I notice that the content is binary in v2 whereas if I use curl I get back a pem file that is ascii as expected.

**Comment by abadger** [comment]
That leads me to suspect there's both a bug in callbacks (should not traceback in v1) and in uri (something changed after 1.9.2 that's causing binary data to be returned in content).

**Comment by abadger** [comment]
Just tried again in 1.9.2... content is also binary that part was not a problem.

**Comment by abadger** [comment]
I think this is fixed by this commit: 73684317ea5b6f9d993256dd9aeff3af05ec10e5
(And some debugging code removed here)189c741ed1706207365007e4d960f5881269e30e

You should be able to replace the uri.py module from 1.9.3 with the latest version of the uri module from either the stable tree: https://github.com/ansible/ansible-modules-core/blob/stable-1.9/network/basics/uri.py

or the devel tree: https://github.com/ansible/ansible-modules-core/blob/devel/network/basics/uri.py to work around this issue until the next release of ansible.

(Just a note: thinking this was being produced by an ansible callback was a red herring.  The callback referenced in the traceback was actually from python about its UnicodeDecode error callback).

# Closing This Ticket

Hi!

We believe recent commits (likely detailed above) should resolve this question or problem for you.

This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
